ring.go

```diff
@@ -220,25 +222,94 @@ type ringShards struct {
 }
 
 func newRingShards(opt *RingOptions) *ringShards {
-	shards := make(map[string]*ringShard, len(opt.Addrs))
+	shards, list := createRingShards(opt, opt.Addrs)
+
+	c := &ringShards{
+		opt: opt,
+
+		shards: shards,
+		list:   list,
+	}
+	c.rebalance()
+
+	return c
+}
+
+func createRingShards(opt *RingOptions, addrs map[string]string) (map[string]*ringShard, []*ringShard) {
+	shards := make(map[string]*ringShard, len(addrs))
 	list := make([]*ringShard, 0, len(shards))
 
-	for name, addr := range opt.Addrs {
+	for name, addr := range addrs {
 		shard := newRingShard(opt, name, addr)
 		shards[name] = shard
 
 		list = append(list, shard)
 	}
+	return shards, list
+}
 
-	c := &ringShards{
-		opt: opt,
+func (c *ringShards) SetAddrs(addrs map[string]string) {
+	var currentRingshards map[string]*ringShard
+	toDelete := make(map[string]*ringShard)
+	newShards := make(map[string]*ringShard)
 
-		shards: shards,
-		list:   list,
+	c.mu.RLock()
+	if !c.closed {
+		currentRingshards = c.shards
+	}
+	c.mu.RUnlock()
+
+	// delete already connected shard addresses
+	for k, currentRingshard := range currentRingshards {
+		if v, ok := addrs[k]; ok {
+
+			if currentRingshard.addr == v {
```

Comment:
The check could be combined into `ok && currentRingshard.addr == v`.
It is also possible to `delete(currentRingshards, k)` right away https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-map-within-a-range-loop

Reply:
Actually I think it is not safe to modify `currentRingshards` as it is just a ref to the `c.shards` (the same is true for `addrs` argument). I think it is better to fill-in the new map+list and then update `c.shards` and `c.list`.

Reply:
I don't want to change the if predicate into one, because there are 2 different else cases, that do the same thing, but maybe there is something we miss.

That's why I don't delete here right away, yes.

Reply:
As far as I understand it should be safe, but and a test with race detector enabled was fine. 
```
ginkgo -race -focus '\[new\] dynamic setting ring shards' .
```
For this to run you need to delete a couple of Examples that panic (also without -race they panic).